portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -22,31 +22,72 @@
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import com.google.appengine.api.datastore.FetchOptions;
 
 /** Servlet that returns some example content. TODO: modify this file to handle comments data */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
   
-  private ArrayList<String> quotes;
+  private List<String> comments = new ArrayList<>(); 
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    quotes = new ArrayList<>();
-    quotes.add("beep boop - ben lin");
-    quotes.add("They told me computers could only do arithmetic -Grace Hopper. I told them I cannot do arithmetic - Ben Lin");
-    quotes.add("hydrate or diedrate - Ben Lin");
-
-    // Convert the quotes to JSON using Gson library
-    String json = new Gson().toJson(quotes);
-
+    
     // Send the JSON as the response
     response.setContentType("application/json;");
-    response.getWriter().println(json);
     
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Query query = new Query("Message").addSort("timestamp", SortDirection.DESCENDING);
+    List<Entity> results = datastore.prepare(query).asList(FetchOptions.Builder.withLimit(10));
+    // add a limit
```

Comment:
nit: implementation comments that explain what code's doing usually goes either above the code it's explaining or to the end:

// add a limit
... FetchOptions.Builder.withLimit(10)
or
... FetchOptions.Builder.withLimit(10) // add a limit

though also try to limit the amount of these comments. In many cases they're redundant (such as this one, it's pretty clear that the query is using a limit), or it's a sign that a block of code can be extracted into a method for clarity. Consider:

// Query from Datastore 
DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
Query query = new Query("Message").addSort("timestamp", SortDirection.DESCENDING);
List<Entity> results = datastore.prepare(query).asList(FetchOptions.Builder.withLimit(10));

vs

List<Entity> results = queryFromDatastore();

private List<Entity> queryFromDatastore() {
  DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
  Query query = new Query("Message").addSort("timestamp", SortDirection.DESCENDING);
  return datastore.prepare(query).asList(FetchOptions.Builder.withLimit(10));
}

The second one helps keep methods focused and self-explanatory, as well as makes code more reusable, as any other method in this class can also call queryFromDatastore() without copying and pasting its logic

Reply:
Great point. Thanks.